mpdev/internal/apply/deployment_manager.go

```diff
@@ -0,0 +1,148 @@
+// given an autogen.yaml file.
+type DeploymentManagerAutogenTemplate struct {
+	ResourceShared
+	AutogenFile string
+	PartnerID   string `json:"partnerId"`
+	SolutionID  string `json:"solutionId"`
+
+	outDir string
+}
+
+// Apply generates a deployment manager template from an autogen file.
+func (dm *DeploymentManagerAutogenTemplate) Apply() error {
+	dir, err := ioutil.TempDir("", "autogen")
+	if err != nil {
+		return err
+	}
+	dm.outDir = dir
+
+	fmt.Printf("Generating deployment manager template from autogen\n")
+	err = dm.convertAutogenSchema()
+	if err != nil {
+		return err
+	}
+
+	autogenImg := "gcr.io/cloud-marketplace-tools/dm/autogen"
+	args := []string{"--input_type", "YAML", "--single_input", "/autogen/autogen.yaml",
+		"--output_type", "PACKAGE", "--output", "/autogen"}
+
+	cp := &ContainerProcess{
+		containerImage: autogenImg,
+		processArgs:    args,
+		mounts:         []string{fmt.Sprintf("type=bind,src=%s,dst=/autogen", dm.outDir)},
+	}
+	cmd := cp.getCommand()
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+
+	err = cmd.Run()
+	if err != nil {
+		return errors.Wrap(err, "failed to execute autogen")
+	}
+
+	fmt.Printf("Wrote autogen output to directory: %s\n", dm.outDir)
+
+	return nil
+}
+
+func (dm *DeploymentManagerAutogenTemplate) convertAutogenSchema() error {
+	inputFile, err := os.Open(dm.AutogenFile)
+	if err != nil {
+		return err
+	}
+	defer inputFile.Close()
+
+	outFile, err := os.Create(filepath.Join(dm.outDir, "autogen.yaml"))
+	if err != nil {
+		return err
+	}
+	defer outFile.Close()
+
+	// Image name from running `docker build mpdev/autogen -f mpdev/autogen/Dockerfile -t autogen_converter
+	autogenConverterImg := "autogen_converter"
+	cp := &ContainerProcess{
+		containerImage: autogenConverterImg,
+		processArgs:    []string{"--partnerId", dm.PartnerID, "--solutionId", dm.SolutionID},
+		mounts:         nil,
+	}
+	cmd := cp.getCommand()
+
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = inputFile
+	cmd.Stdout = outFile
+
+	err = cmd.Run()
+	if err != nil {
+		return errors.Wrap(err, "failed to execute autogen converter")
+	}
+
+	return nil
+}
+
+// DeploymentManagerTemplateOnGCS uploads a referenced Deployment Manager
+// template to GCS.
+type DeploymentManagerTemplateOnGCS struct {
+	ResourceShared
+	DeploymentManagerRef Reference
+	GCS                  struct {
+		Bucket string
+		Object string
+	}
+}
+
+// Apply uploads a Deployment Manager template to GCS.
+func (dm *DeploymentManagerTemplateOnGCS) Apply() error {
+	dmRef := dm.referenceMap[dm.DeploymentManagerRef]
+	if dmRef == nil {
+		return fmt.Errorf("autogen template not found %+v", dm.DeploymentManagerRef)
+	}
+
+	dmTemplate, ok := dmRef.(*DeploymentManagerAutogenTemplate)
+	if !ok {
+		return fmt.Errorf("referenced autogen template is not correct type %+v", dm.DeploymentManagerRef)
+	}
+
+	gcsPath := fmt.Sprintf("gs://%s/%s", dm.GCS.Bucket, dm.GCS.Object)
+
+	cmd := exec.Command("gsutil", "cp", "-r", dmTemplate.outDir, gcsPath)
```

Comment:
we need to zip the files. That could be a default option in the resource configuration

Reply:
Done